src/System.Runtime.Numerics/src/System/Numerics/BigCalc.SquMul.cs

```diff
@@ -0,0 +1,356 @@
+﻿using System.Diagnostics;
+using System.Security;
+
+namespace System.Numerics
+{
+    internal static partial class BigCalc
+    {
+#if DEBUG
+
+        private const int SquareThreshold = 128 / 32;
+        private const int MultiplyThreshold = 128 / 32;
+
+#else
+
+        private const int SquareThreshold = 1024 / 32;
+        private const int MultiplyThreshold = 1024 / 32;
+
+#endif
+
+        public static uint[] Square(uint[] value)
+        {
+            Debug.Assert(value != null);
+
+            return Square(value, value.Length);
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static uint[] Square(uint[] value, int valueLength)
```

Comment:
But there are calls where `valueLength` is smaller than `value.Length` (after the "fold" part within *square*). And we're talking of a *private* method.